dcos_test_utils/ssh_client.py

```diff
@@ -136,3 +132,133 @@ def wait_for_ssh_connection(self, host: str, port: int=22) -> None:
+                stderr=asyncio.subprocess.PIPE,
+                stdin=slave_pty,
+                env={'TERM': 'linux'})
+            stdout = b''
+            stderr = b''
+            try:
+                stdout, stderr = yield from asyncio.wait_for(process.communicate(), self.process_timeout)
+            except asyncio.TimeoutError:
+                try:
+                    process.terminate()
+                except ProcessLookupError:
+                    log.info('process with pid {} not found'.format(process.pid))
+                log.error('timeout of {} sec reached. PID {} killed'.format(self.process_timeout, process.pid))
+
+        return {
+            "cmd": cmd,
+            "stdout": stdout,
+            "stderr": stderr,
+            "returncode": process.returncode,
+            "pid": process.pid
+        }
+
+    @asyncio.coroutine
+    def run_async(self, host: str, cmd: list) -> dict:
+        """ runs cmd as arguments on host
+        """
+        hostname, port = parse_ip(host)
+        with self.tunnel(hostname, port) as t:
+            full_cmd = t.base_cmd + [t.target] + cmd
+            log.debug('executing command {}'.format(full_cmd))
+            result = yield from self.run_cmd_return_dict_async(full_cmd)
+        result['host'] = host
+        return result
+
+    @asyncio.coroutine
+    def copy_async(self, host: str, local_path: str, remote_path: str, recursive: bool) -> dict:
+        """ instructs SCP to copy local_path to remote_path on host. recursive copying may be enabled
+        """
+        hostname, port = parse_ip(host)
+        copy_command = []
+        if recursive:
+            copy_command.append('-r')
+        remote_full_path = '{}@{}:{}'.format(self.user, hostname, remote_path)
+        copy_command += [local_path, remote_full_path]
+        full_cmd = ['/usr/bin/scp'] + SHARED_SSH_OPTS + ['-P', str(port), '-i', self.key_path] + copy_command
+        log.debug('copy with command {}'.format(full_cmd))
+        result = yield from self.run_cmd_return_dict_async(full_cmd)
+        result['host'] = host
+        return result
+
+    @asyncio.coroutine
+    def run_command_on_hosts(self, coroutine_name: str, *args) -> list:
+        """ coroutine to run a single coroutine against all hosts in the MultiRunner
+        """
+        tasks = yield from self.start_command_on_hosts(coroutine_name, *args)
+        yield from asyncio.wait(tasks)
+        return [task.result() for task in tasks]
+
+    @asyncio.coroutine
+    def start_command_on_hosts(self, coroutine_name: str, *args, sem: asyncio.Semaphore=None) -> list:
+        """ coroutine to run a single coroutine against all hosts in the MultiRunner
+        """
+        if sem is None:
+            sem = asyncio.Semaphore(self.__parallelism)
+        log.debug('Waiting for run_command_chain_async to execute')
+        tasks = []
+        for host in self.__targets:
+            with (yield from sem):
```

Comment:
Ouch at this python syntax of `with (yield from semaphore)`

Reply:
@orsenthil I am not a huge fan of this syntax either. Seems like black magic to me. Will change this so that the individual coroutines acquire within the method